### PR TITLE
refactor(Companion): Use QHttp server

### LINKED ISSUE #345: Competitive companion parser error
While parsing problems from competitve companion a JSON parser error reflects. It resolves on restart cp editor. But it is irritating.
Screenshot from 2020-05-13 15-39-32

 - OS: Ubuntu 18.04
 - CP Editor Version: 6.2.5

### PR DESCRIPTION
**Pull request by coder3101**
## Related Issue
Fixes #345 (Confirmation required)

## Motivation and Context
It is reported that in some cases the Competitive companion server is skipping some data or not at all accepting valid data on Chrome browser. This could not be reprodduced by us, but a possible solution was to use some other ready made Qt Http Server. In this PR, We use that server to listen for requests sent by Competitive Companion. This should solve the issue. (confirmation required)

## How Has This Been Tested?
Arch Linux with KDE

## Screenshots (if appropriate)
N/A
## Type of changes

- [x] Bug fix (changes which fix an issue)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/cpeditor/cpeditor/blob/master/CONTRIBUTING.md) document.
- [x] I have tested these changes locally, and this fixes the bug/the new feature behaves as the expectation.
- [x] The settings file in the old version can be used in the new version after this change.
- [x] These changes only fix a single bug/introduces a single feature. (Otherwise, open multiple Pull Requests instead, unless these bugs/features are closely related.)
- [x] The commit messages are clear and detailed. (Otherwise, use `git reset` and commit again, or use `git rebase -i` and `git commit --amend` to modify the commit messages.)
- [x] These changes don't remove an existing feature. (Otherwise, add an option to disable this feature instead, unless it's necessary to remove this feature.)
- [x] If there are changes of the text displayed in the UI, I have wrapped them in `tr()` or `QCoreApplication::translate()`.
- [x] I have documented these changes in [CHANGELOG.md](https://github.com/cpeditor/cpeditor/blob/master/CHANGELOG.md), or these changes are not notable.

### CODE REVIEW

**Comment by coder3101** [comment]
S*** Windows MSVC Compiler is failing. Will fix it when on Windows!

**Comment by coder3101** [comment]
@IZOBRETATEL777 and @ouuan translations required.

File: src/Extensions/CompanionServer.cpp [hunk 1]
```diff
@@ -18,148 +18,146 @@
...
+                           res->addHeader("connection", "close");
+                           res->addHeader("pragma", "no-cache");
+
+                           if (methodType != "POST")
+                           {
+                               SAFEWLOG("Companion", tr("Discarded %1 request").arg(methodType));
+                               res->setStatusCode(qhttp::ESTATUS_NOT_ACCEPTABLE);
+                               res->end();
+                               return;
+                           }
```
Comment (ouuan) [inline src/Extensions/CompanionServer.cpp:78]: What does this error mean? When does it happen? Why is it a warning?

**Comment by ouuan** [comment]
Is it guaranteed that the `content-type` is not other words like `Content Type` on all platforms?

**Comment by coder3101** [comment]
> Is it guaranteed that the `content-type` is not other words like `Content Type` on all platforms?

Yes as per QHttpServer documentation, the header is always lowercase. Moreover, the function `keyHasValue(a,b)` compares the key and value in case insensitive matter.

File: src/Extensions/CompanionServer.cpp (hunk 1 above)
Comment (coder3101) [reply src/Extensions/CompanionServer.cpp]: When a GET/PUT/DELETE or any other non POST request is made to the server port. We Gracefully ignore it and issue a warning. We could have ignored it all together even without logging a warning but if this happens user should know that something else in their computer is making some request to the Port which is suppose to listen for only Competitive Companion POST data.

Comment (coder3101) [reply src/Extensions/CompanionServer.cpp]: One way to trigger this warning is that you could visit `localhost:10045` (port of competitive companion) from your web browser. It sends a GET request and CP Editor will issue the warning. From development point of view it is another way to check if server is actually listening on the port.

File: src/Extensions/CompanionServer.cpp [hunk 2]
```diff
@@ -18,148 +18,143 @@
...
+    server->listen(QString::number(port), [this](qhttp::server::QHttpRequest *req, qhttp::server::QHttpResponse *res) {
+        LOG_INFO("\n--> " << req->methodString() << " : " << qPrintable(req->url().toString().toUtf8()));
+        req->headers().forEach(
+            [](auto iter) { LOG_INFO(iter.key().constData() << " : " << iter.value().constData()); });
+
+        const QString methodType = req->methodString();
+        const bool isJson = req->headers().keyHasValue("content-type", "application/json");
+        req->collectData();
+
+        req->onEnd([=] {
```
Comment (coder3101) [inline src/Extensions/CompanionServer.cpp:67]: Why not capture by reference?

**Comment by coder3101** [comment]
@IZOBRETATEL777 or @sadykhzadeh Could you please add Russian translation? 

A translation of these two statements is required.

1. "The request received is not JSON"
2. "A %1 request is received and ignored"

If it is not possible for you to sync this PR. You could atleast reply to this thread with the translations for the above statement.

File: src/Extensions/CompanionServer.cpp (hunk 2 above)
Comment (ouuan) [reply src/Extensions/CompanionServer.cpp:67]: I think they are local variables and will be destroyed at the end of `CompanionServer::startListeningOn`, which results in undefined behavior.

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
 ### Fixed
 
--   We now use QHttp to listen for Competitive Companion data. This should fix an issue where in some cases CP Editor could not get full data from Google Chrome browsers. (#345 and #624)
+-   Fixed an issue where in some cases CP Editor could not get full data from Google Chrome browsers. (#345 and #624)
```
Comment (ouuan) [inline CHANGELOG.md:7]:
```suggestion
-   Fixed an issue that in some cases CP Editor could not get full data from Competitive Companion. (#345 and #624)
```

We have no evidence that it only happens in Chrome.

**Review by ouuan — APPROVED** [review]

**Pull request merged**